gateway/ln-gateway/src/lnd.rs

```diff
@@ -0,0 +1,302 @@
+    ) -> crate::Result<Self> {
+        let client = connect(address, tls_cert, macaroon).await.map_err(|e| {
+            error!("Failed to connect to lnrpc server: {:?}", e);
+            GatewayError::Other(anyhow!("Failed to connect to lnrpc server"))
+        })?;
+
+        Ok(Self {
+            client,
+            outcomes: Arc::new(Mutex::new(HashMap::new())),
+            task_group,
+        })
+    }
+}
+
+impl fmt::Debug for GatewayLndClient {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "LndClient")
+    }
+}
+
+#[async_trait]
+impl ILnRpcClient for GatewayLndClient {
+    async fn pubkey(&self) -> crate::Result<GetPubKeyResponse> {
+        let mut client = self.client.clone();
+
+        let info = client
+            .lightning()
+            .get_info(GetInfoRequest {})
+            .await
+            .expect("failed to get info")
+            .into_inner();
+
+        let pub_key: PublicKey = info.identity_pubkey.parse().expect("invalid pubkey");
+        info!("fetched pubkey {:?}", pub_key);
+
+        Ok(GetPubKeyResponse {
+            pub_key: pub_key.serialize().to_vec(),
+        })
+    }
+
+    async fn routehints(&self) -> crate::Result<GetRouteHintsResponse> {
+        Ok(GetRouteHintsResponse {
+            route_hints: vec![RouteHint { hops: vec![] }],
+        })
+    }
+
+    async fn pay(&self, invoice: PayInvoiceRequest) -> crate::Result<PayInvoiceResponse> {
+        let mut client = self.client.clone();
+
+        let send_response = client
+            .lightning()
+            .send_payment_sync(SendRequest {
+                payment_request: invoice.invoice.to_string(),
+                ..Default::default()
+            })
+            .await
+            .map_err(|e| anyhow::anyhow!(format!("LND error: {e:?}")))?
+            .into_inner();
+        info!("send response {:?}", send_response);
+
+        if send_response.payment_preimage.is_empty() {
+            return Err(GatewayError::LnRpcError(tonic::Status::new(
+                tonic::Code::Internal,
+                "LND did not return a preimage",
+            )));
+        };
+
+        Ok(PayInvoiceResponse {
+            preimage: send_response.payment_preimage,
+        })
+    }
+
+    async fn subscribe_htlcs<'a>(
+        &self,
+        subscription: SubscribeInterceptHtlcsRequest,
+    ) -> crate::Result<HtlcStream<'a>> {
+        let channel_size = 1024;
+
+        // Channel to send responses to LND after processing intercepted HTLC
+        let (lnd_tx, lnd_rx) = mpsc::channel::<ForwardHtlcInterceptResponse>(channel_size);
+
+        // Channel to send intercepted htlc to gatewayd for processing
+        let (gwd_tx, gwd_rx) =
+            mpsc::channel::<Result<SubscribeInterceptHtlcsResponse, tonic::Status>>(channel_size);
+
+        let scid = subscription.short_channel_id;
+        let mut client = self.client.clone();
+
+        let mut tg = self.task_group.clone();
```

Comment:
If I recall correctly, isn't there a way to make a subgroup instead of cloning the original group?

Reply:
I think this is actually as good as we can do here. This task group is just to spawn the htlc subscription task, so we shouldn't need another child.

I made another commit that makes sure that `GatewayLndClient::new` always gets a new subgroup, and added a comment about the purpose of this task group.

While cloning sucks, this basically just cloning an `Arc` so it's pretty cheap ...